core/controllers/platform_feature.py

```diff
@@ -29,16 +29,39 @@ class PlatformFeaturesEvaluationHandler(base.BaseHandler):
     """The handler for retrieving feature flag values."""
 
     GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+    URL_PATH_ARGS_SCHEMAS = {}
+    HANDLER_ARGS_SCHEMAS = {
```

Comment:
How do we ensure that these schemas are correct? Should any new tests be added, or are these somehow generically covered in platform-wide tests?

Reply:
True they are covered in platform wide test.
Although we have test_cases (backend & linting) to ensure the structure of schema is correct and the behaviour is tested by existing tests of handlers.

Reply:
Hmm, but there aren't any tests verifying that the schema checks fail in the expected cases (so far as I'm aware). Shouldn't we be adding those?

Reply:
@BenHenning The existing tests in the controller test files performs tests with different valid and invalid payloads and thus the schema validation is actually being tested and there is no need to add additional tests. Whereas the overall working of the schema validation and the correctness of the schema is tested in the base_test file. So I don't think there a need for adding additional tests.

Reply:
Thank you @rohitkatlaa.

Reply:
@rohitkatlaa can you point me to the specific tests? From what I can see, the tests verify that different platform types result in specific behaviors (unrelated to the schema, though we aren't testing non-strings which would trigger the schema failure), and a partly invalid app version (which interestingly doesn't fail the schema check--do you know why?).

I don't see anything validating that browser type is correct, or ensuring that different implied behaviors of the regex check actually work as expected (i.e. that there's a hash, a properly formatted major/minor version, etc.).

To this end, there are no tests that fail if we remove these checks which technically means we aren't covering them in regression testing (unless I'm missing something), and if we want to ensure these checks stay in place we should probably have regression tests to verify them fully.

Reply:
I agree with @BenHenning that tests have to added to verify the working of the schema, there are missing tests for the browser_type. And regarding the partly invalid app version, it passes the schema check because it matches the regex, but it actually fails because the flavor is not in the choices(ALLOWED_APP_VERSION_FLAVORS) which is not being checked by the schema validation. So please @Nik-09 add the additional validation and the tests that check with invalid payloads.

Reply:
Thank you I have added a few test cases and in this thread I have included some explanation, please take a look.
By reading your message I understood the requirement of adding test cases which cover mainly these points.
1. Adding tests for non-strings-platform types.
    - Since platform parameter type is a url query param, it is always string. If I try to send True, then it will be converted into ‘True’.
    - This is exactly the same case for Invalid, thus no need to add any extra case for platform types.
2. Adding tests for invalid app version
    - The existing regex for checking app-version are of two types (a). With hash (b). Without hash.
        - The regex PLATFORM_PARAMETER_APP_VERSION_WITH_HASH_REGEXP takes care of the fact that if the app-version is without hash then also it can pass as far as the regex of the version is correct.
    - Added test to verify the regex with an invalid app-version.
    - For app-version-flavours we can’t put choices since flavours are a part of the version not the entire argument and also I think this is an optional part of the version.
3. Adding tests for invalid browser type
    - I have added a test case to verify the browser type that invalid one should raise 400.

Please correct me if I am wrong anywhere.
Thank you.

Reply:
The new tests seem good--thanks @Nik-09 & @rohitkatlaa.